store/postgres/migrations/2018-08-22-140000_create_head_update_procedure/up.sql

```diff
@@ -0,0 +1,76 @@
+/**************************************************************
+* CREATE BLOCK INGESTOR FUNCTIONS
+**************************************************************/
+
+/**************************************************************
+* Attempt to update the head block pointer.
+* Raises an exception if ethereum_blocks table is empty.
+**************************************************************/
+CREATE OR REPLACE FUNCTION attempt_head_update(net_name VARCHAR, ancestor_count BIGINT)
+    RETURNS VARCHAR[] AS
+$$
+DECLARE
+    current_head_number BIGINT;
+    new_head_hash VARCHAR;
+    new_head_number BIGINT;
+    missing_parents VARCHAR[];
+BEGIN
+    -- Get block number of current head block
+    SELECT head_block_number
+    INTO STRICT current_head_number
+    FROM ethereum_networks
+    WHERE name = net_name;
+
+    -- Find candidate new head block
+    SELECT
+        hash,
+        number
+    INTO STRICT
+        new_head_hash,
+        new_head_number
+    FROM ethereum_blocks
+    WHERE network_name = net_name
+    ORDER BY
+        number DESC,
+        hash ASC
+    LIMIT 1;
+
+    -- Stop now if it's no better than the current head block
+    IF new_head_number <= current_head_number THEN
+        RETURN ARRAY[]::VARCHAR[];
+    END IF;
+
+    -- Aggregate list of missing parent hashes into missing_parents,
+    -- selecting only parents of blocks within ancestor_count of new head
+    SELECT array_agg(block1.parent_hash)
+    INTO STRICT missing_parents
+    FROM ethereum_blocks AS block1
+    LEFT OUTER JOIN ethereum_blocks AS block2
+    ON block1.parent_hash = block2.hash
+    WHERE
+        block1.network_name = net_name
+        AND block2.hash IS NULL -- cases where no block2 was found
+        AND block1.number > (new_head_number - ancestor_count);
+
+    -- Stop now if there are any recent blocks with missing parents
+    IF array_length(missing_parents, 1) > 0 THEN
+        RETURN missing_parents;
+    END IF;
+
+    -- No recent missing parent blocks, therefore candidate new head block has
+    -- the necessary minimum number of ancestors present in DB.
+
+    -- Set head block pointer to candidate head block
+    UPDATE ethereum_networks
+    SET
+        head_block_hash = new_head_hash,
+        head_block_number = new_head_number
+    WHERE name = net_name;
+
+    -- Fire head block update event
+    NOTIFY head_block_update;
```

Comment:
This works with a single network but it would be better to include the network name and the new head block hash and number in the notification here (e.g. with `pg_notify` + `json_build_object`).

Reply:
I was looking to do that originally, but surprisingly the payload has to be a constant string in postgres: https://www.postgresql.org/docs/9.6/static/sql-notify.html

I could workaround that, but I think it's OK -- checking if the head block has changed is quick, and these extra checks would only happen if processing is already caught up

Reply:
(maybe worth putting in a separate issue to investigate as an optimization?)

Reply:
It does take a constant string, but that's only for the channel name. It has a second, optional `payload` parameter. See here ;)

https://github.com/graphprotocol/graph-node/blob/master/store/postgres/migrations/2018-08-16-143800_create_entity_change_triggers/up.sql#L13

Reply:
`pg_notify` did the trick. Done in the tim/block-processing branch: https://github.com/graphprotocol/graph-node/commit/9a7fc2f1df8815a017fda5f7a104732e0b78b590#diff-e69c79fa22704b08685b3412c3ba9eabR75